File: benchmarks/n-body/c/nbody.c
/*
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * N-body simulation of the Jovian planets.
 */

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define PI 3.141592653589793
#define SOLAR_MASS (4 * PI * PI)
#define DAYS_PER_YEAR 365.24

#define NBODIES 5

typedef struct {
    double x, y, z;
    double vx, vy, vz;
    double mass;
} Body;

static Body bodies[NBODIES] = {
    /* Sun */
    { 0, 0, 0, 0, 0, 0, SOLAR_MASS },
    /* Jupiter */
    {
         4.84143144246472090e+00,
        -1.16032004402742839e+00,
        -1.03622044471123109e-01,
         1.66007664274403694e-03 * DAYS_PER_YEAR,
         7.69901118419740425e-03 * DAYS_PER_YEAR,
        -6.90460016972063023e-05 * DAYS_PER_YEAR,
         9.54791938424326609e-04 * SOLAR_MASS
    },
    /* Saturn */
    {
         8.34336671824457987e+00,
         4.12479856412430479e+00,
        -4.03523417114321381e-01,
        -2.76742510726862411e-03 * DAYS_PER_YEAR,
         4.99852801234917238e-03 * DAYS_PER_YEAR,
         2.30417297573763929e-05 * DAYS_PER_YEAR,
         2.85885980666130812e-04 * SOLAR_MASS
    },
    /* Uranus */
    {
         1.28943695621391310e+01,
        -1.51111514016986312e+01,
        -2.23307578892655734e-01,
         2.96460137564761618e-03 * DAYS_PER_YEAR,
         2.37847173959480950e-03 * DAYS_PER_YEAR,
        -2.96589568540237556e-05 * DAYS_PER_YEAR,
         4.36624404335156298e-05 * SOLAR_MASS
    },
    /* Neptune */
    {
         1.53796971148509165e+01,
        -2.59193146099879641e+01,
         1.79258772950371181e-01,
         2.68067772490389322e-03 * DAYS_PER_YEAR,
         1.62824170038242295e-03 * DAYS_PER_YEAR,
        -9.51592254519715870e-05 * DAYS_PER_YEAR,
         5.15138902046611451e-05 * SOLAR_MASS
    }
};

static void offset_momentum(void) {
    double px = 0, py = 0, pz = 0;
    for (int i = 0; i < NBODIES; i++) {
        px += bodies[i].vx * bodies[i].mass;
        py += bodies[i].vy * bodies[i].mass;
        pz += bodies[i].vz * bodies[i].mass;
    }
    bodies[0].vx = -px / SOLAR_MASS;
    bodies[0].vy = -py / SOLAR_MASS;
    bodies[0].vz = -pz / SOLAR_MASS;
}

static double energy(void) {
    double e = 0.0;
    for (int i = 0; i < NBODIES; i++) {
        Body *bi = &bodies[i];
        e += 0.5 * bi->mass * (bi->vx * bi->vx + bi->vy * bi->vy + bi->vz * bi->vz);
        for (int j = i + 1; j < NBODIES; j++) {
            Body *bj = &bodies[j];
            double dx = bi->x - bj->x;
            double dy = bi->y - bj->y;
            double dz = bi->z - bj->z;
            double dist = sqrt(dx * dx + dy * dy + dz * dz);
            e -= (bi->mass * bj->mass) / dist;
        }
    }
    return e;
}

static void advance(double dt) {
    for (int i = 0; i < NBODIES; i++) {
        Body *bi = &bodies[i];
        for (int j = i + 1; j < NBODIES; j++) {
            Body *bj = &bodies[j];
            double dx = bi->x - bj->x;
            double dy = bi->y - bj->y;
            double dz = bi->z - bj->z;
            double dsq = dx * dx + dy * dy + dz * dz;
            double dist = sqrt(dsq);
            double mag = dt / (dsq * dist);
            bi->vx -= dx * bj->mass * mag;
            bi->vy -= dy * bj->mass * mag;
            bi->vz -= dz * bj->mass * mag;
            bj->vx += dx * bi->mass * mag;
            bj->vy += dy * bi->mass * mag;
            bj->vz += dz * bi->mass * mag;
        }
    }
    for (int i = 0; i < NBODIES; i++) {
        bodies[i].x += dt * bodies[i].vx;
        bodies[i].y += dt * bodies[i].vy;
        bodies[i].z += dt * bodies[i].vz;
    }
}

int main(int argc, char *argv[]) {
    int n = atoi(argv[1]);
    offset_momentum();
    printf("%.9f\n", energy());
    for (int i = 0; i < n; i++)
        advance(0.01);
    printf("%.9f\n", energy());
    return 0;
}

File: benchmarks/n-body/java/NBody.java
/**
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * N-body simulation of the Jovian planets.
 */
public final class NBody {

    static final double PI = Math.PI;
    static final double SOLAR_MASS = 4 * PI * PI;
    static final double DAYS_PER_YEAR = 365.24;

    static final double[] x  = new double[5];
    static final double[] y  = new double[5];
    static final double[] z  = new double[5];
    static final double[] vx = new double[5];
    static final double[] vy = new double[5];
    static final double[] vz = new double[5];
    static final double[] mass = new double[5];

    static {
        // Sun
        x[0]=0; y[0]=0; z[0]=0; vx[0]=0; vy[0]=0; vz[0]=0;
        mass[0] = SOLAR_MASS;
        // Jupiter
        x[1]=4.84143144246472090e+00;
        y[1]=-1.16032004402742839e+00;
        z[1]=-1.03622044471123109e-01;
        vx[1]=1.66007664274403694e-03*DAYS_PER_YEAR;
        vy[1]=7.69901118419740425e-03*DAYS_PER_YEAR;
        vz[1]=-6.90460016972063023e-05*DAYS_PER_YEAR;
        mass[1]=9.54791938424326609e-04*SOLAR_MASS;
        // Saturn
        x[2]=8.34336671824457987e+00;
        y[2]=4.12479856412430479e+00;
        z[2]=-4.03523417114321381e-01;
        vx[2]=-2.76742510726862411e-03*DAYS_PER_YEAR;
        vy[2]=4.99852801234917238e-03*DAYS_PER_YEAR;
        vz[2]=2.30417297573763929e-05*DAYS_PER_YEAR;
        mass[2]=2.85885980666130812e-04*SOLAR_MASS;
        // Uranus
        x[3]=1.28943695621391310e+01;
        y[3]=-1.51111514016986312e+01;
        z[3]=-2.23307578892655734e-01;
        vx[3]=2.96460137564761618e-03*DAYS_PER_YEAR;
        vy[3]=2.37847173959480950e-03*DAYS_PER_YEAR;
        vz[3]=-2.96589568540237556e-05*DAYS_PER_YEAR;
        mass[3]=4.36624404335156298e-05*SOLAR_MASS;
        // Neptune
        x[4]=1.53796971148509165e+01;
        y[4]=-2.59193146099879641e+01;
        z[4]=1.79258772950371181e-01;
        vx[4]=2.68067772490389322e-03*DAYS_PER_YEAR;
        vy[4]=1.62824170038242295e-03*DAYS_PER_YEAR;
        vz[4]=-9.51592254519715870e-05*DAYS_PER_YEAR;
        mass[4]=5.15138902046611451e-05*SOLAR_MASS;
    }

    static void offsetMomentum() {
        double px = 0, py = 0, pz = 0;
        for (int i = 0; i < 5; i++) {
            px += vx[i] * mass[i];
            py += vy[i] * mass[i];
            pz += vz[i] * mass[i];
        }
        vx[0] = -px / SOLAR_MASS;
        vy[0] = -py / SOLAR_MASS;
        vz[0] = -pz / SOLAR_MASS;
    }

    static double energy() {
        double e = 0.0;
        for (int i = 0; i < 5; i++) {
            e += 0.5 * mass[i] * (vx[i]*vx[i] + vy[i]*vy[i] + vz[i]*vz[i]);
            for (int j = i + 1; j < 5; j++) {
                double dx = x[i] - x[j];
                double dy = y[i] - y[j];
                double dz = z[i] - z[j];
                double dist = Math.sqrt(dx*dx + dy*dy + dz*dz);
                e -= (mass[i] * mass[j]) / dist;
            }
        }
        return e;
    }

    static void advance(double dt) {
        for (int i = 0; i < 5; i++) {
            for (int j = i + 1; j < 5; j++) {
                double dx = x[i] - x[j];
                double dy = y[i] - y[j];
                double dz = z[i] - z[j];
                double dsq = dx*dx + dy*dy + dz*dz;
                double dist = Math.sqrt(dsq);
                double mag = dt / (dsq * dist);
                vx[i] -= dx * mass[j] * mag;
                vy[i] -= dy * mass[j] * mag;
                vz[i] -= dz * mass[j] * mag;
                vx[j] += dx * mass[i] * mag;
                vy[j] += dy * mass[i] * mag;
                vz[j] += dz * mass[i] * mag;
            }
        }
        for (int i = 0; i < 5; i++) {
            x[i] += dt * vx[i];
            y[i] += dt * vy[i];
            z[i] += dt * vz[i];
        }
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        offsetMomentum();
        System.out.printf("%.9f\n", energy());
        for (int i = 0; i < n; i++) {
            advance(0.01);
        }
        System.out.printf("%.9f\n", energy());
    }
}
